src/main/scala/in/ashwanthkumar/suuchi/router/ReplicationRouter.scala

```diff
@@ -66,36 +69,90 @@ abstract class ReplicationRouter(nrReplicas: Int, self: MemberAddress) extends S
-        }
-        // we need to push this after the forwarding else we return to client immediately saying we're done
-        if (iamfound) {
-          delegate.onMessage(incomingRequest)
-          forwarded = false // FIXME - we had to unset this here since the actual service method is invoked only on onHalfClose()
-        }
+        doReplication(eligibleNodes, serverCall, headers, incomingRequest, delegate)
       case Nil =>
         log.error("This should never happen. No nodes found to place replica")
         serverCall.close(Status.INTERNAL, headers)
     }
   }
+
+  /**
+   * Implement the actual replication logic assuming that you've the right set of nodes.
+   * Just do it!
+   *
+   * Error handling and other scenarios are handled at [[ReplicationRouter.replicate]]
+   * */
+  def doReplication[ReqT, RespT](eligibleNodes: List[MemberAddress], serverCall: ServerCall[ReqT, RespT], headers: Metadata, incomingRequest: ReqT, delegate: Listener[ReqT]): Unit
+}
+
+class SequentialReplicator(nrReplicas: Int, self: MemberAddress) extends ReplicationRouter(nrReplicas, self) {
+  override def doReplication[ReqT, RespT](eligibleNodes: List[MemberAddress], serverCall: ServerCall[ReqT, RespT], headers: Metadata, incomingRequest: ReqT, delegate: Listener[ReqT]) = {
+    log.debug("Sequentially sending out replication requests to the above set of nodes")
+
+    val hasLocalMember = eligibleNodes.exists(_.equals(self))
+
+    eligibleNodes.filterNot(_.equals(self)).foreach { destination =>
+      forward(serverCall.getMethodDescriptor, headers, incomingRequest, destination)
+    }
+
+    // we need to push this after the forwarding else we return to client immediately saying we're done
+    if(hasLocalMember) {
+      delegate.onMessage(incomingRequest)
+      forwarded = false // FIXME - we had to unset this here since the actual service method is invoked only on onHalfClose()
+    }
+  }
+}
+
+object ParallelReplicator {
+  implicit val PARALLEL_REPLICATION_EXECUTOR = Executors.newFixedThreadPool(3)
+}
+class ParallelReplicator(nrReplicas: Int, self: MemberAddress) extends ReplicationRouter(nrReplicas, self) {
+  import ParallelReplicator._
+  override def doReplication[ReqT, RespT](eligibleNodes: List[MemberAddress], serverCall: ServerCall[ReqT, RespT], headers: Metadata, incomingRequest: ReqT, delegate: Listener[ReqT]): Unit = {
+    log.debug("Sending out replication requests to the above set of nodes in parallel")
+
+    val hasLocalMember = eligibleNodes.exists(_.equals(self))
+
+    val replicationResponses = eligibleNodes.filterNot(_.equals(self)).map { destination =>
+        forwardAsync(serverCall.getMethodDescriptor, headers, incomingRequest, destination)
+    }
+
+    // Future.sequence equivalent + doing a get to ensure all operations complete
+    log.debug("Waiting for replication response from replica nodes")
+    Futures.allAsList(replicationResponses:_*).get()
```

Comment:
The returned `Future` fails or cancels immediately if any of the `Future` in the list fails or gets cancelled. So, I guess, it would bomb & throw an exception, which means, the call would return to the client as a failure, which is what we want.

http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/util/concurrent/Futures.html#allAsList(java.lang.Iterable)

In all honesty, we have not done any sort of error handling across the board. Need to look at it soon.